lib/pbench/test/unit/server/database/test_server_config_db.py

```diff
@@ -1,56 +1,309 @@
+        """
+        self.id = 1
+        self.added: List[ServerConfig] = []
+        self.known: Dict[int, ServerConfig] = {}
+        self.committed: Dict[int, FakeRow] = {}
+        self.rolledback = 0
+        self.queries: List[FakeQuery] = []
+        self.raise_on_commit: Optional[Exception] = None
+
+    def query(self, *entities, **kwargs) -> FakeQuery:
+        """
+        Perform a mocked query on the session, setting up the query context
+        and returning it
+
+        Args:
+            entities: The SQLAlchemy entities on which we're operating
+            kwargs: Additional SQLAlchemy parameters
+
+        Returns:
+            A mocked query object
+        """
+        q = FakeQuery(self)
+        self.queries.append(q)
+        return q
+
+    def add(self, config: ServerConfig):
+        """
+        Add a DB object to a list for testing
+
+        Args:
+            config: A server configuration setting object
+        """
+        self.added.append(config)
+
+    def commit(self):
+        """
+        Mock a commit operation on the DB session. If the 'raise_on_commit'
+        property has been set, "fail" by raising the exception. Otherwise,
+        mock a commit by updating any cached "committed" values if the "known"
+        proxy objects have changed, and record any new "added" objects.
+        """
+        if self.raise_on_commit:
+            raise self.raise_on_commit
+        for k in self.known:
+            if self.committed[k].value != self.known[k].value:
+                self.committed[k].value = self.known[k].value
+        for a in self.added:
+            a.id = self.id
+            self.id += 1
+            self.known[a.id] = a
+            self.committed[a.id] = FakeRow(id=a.id, key=a.key, value=a.value)
+        self.added = []
+        self.deleted = []
+
+    def rollback(self):
+        """
+        Just record that rollback was called, since we always raise an error
+        before changing anything during the mocked commit.
+        """
+        self.rolledback += 1
+
+
 class TestServerConfig:
-    def test_bad_key(self, db_session):
+    session = None
+
+    def check_session(self, added=[], queries=0, committed=[], rolledback=0):
+        """
+        Encapsulate the common checks we want to make after running test cases.
+
+        Args:
+            added: A list of object proxies we expect to have been added (but
+                not yet committed)
+            queries: A count of queries we expect to have been made
+            committed: A set of objects we expect to have been committed
+            rolledback: True if we expect rollback to have been called
+        """
+        session = self.session
+        assert session
+        assert len(session.queries) == queries
+        assert len(session.added) == len(added)
+        xlated = [
+            a
+            for a in session.added
+            for k in added
+            if a.key == k.key and a.value == k.value
+        ]
```

Comment:
I'm confused:  this isn't so much "translated" as "filtered"...which is fine if it's appropriate, but seems like we could produce a false positive by throwing stuff away that we thought we added but didn't (but, perhaps I misunderstand).

I think we want to make sure that everything that got in was something we added as well as making sure that everything that we added actually got in, and I _think_ we're only doing the former.

Reply:
And ... even bettererer, with the `rollback` change to wipe out `session.added` this no longer matters anyway since there's no "interesting" visible state where `session.added` isn't empty.